src/jit/hwintrinsicxarch.cpp

```diff
@@ -685,6 +685,171 @@ static bool impIsTableDrivenHWIntrinsic(NamedIntrinsic intrinsicId, HWIntrinsicC
            HWIntrinsicInfo::RequiresCodegen(intrinsicId) && !HWIntrinsicInfo::HasSpecialImport(intrinsicId);
 }
 
+//----------------------------------------------------------------------------
+// isImmExprConstCandidate   Method tries to check passed expression if result
+//                           of evaluation could be an integral const. It limits
+//                           supported node types to explicit list of arithmetic,
+//                           logical, comparison and cast nodes and accepts only
+//                           integral types.
+//
+// Arguments:
+//     tree   -- expression tree node to analyze
+//     isRoot -- true if passed tree is a root of analyzed expression
+//
+// Return value:
+//     returns true if expression is good constant folding candidate otherwise false
+//
+bool Compiler::isImmExprConstCandidate(GenTree* tree, bool isRoot /* = false */)
+{
+    if (!opts.OptEnabled(CLFLG_CONSTANTFOLD))
+    {
+        return false;
+    }
+
+    if (optValnumCSE_phase)
+    {
+        return false;
+    }
+
+    assert(tree != nullptr);
+
+    if (!varTypeIsIntegral(tree->gtType))
+    {
+        return false;
+    }
+
+    genTreeOps oper = tree->OperGet();
+    unsigned   kind = tree->OperKind();
+
+    if (!(kind & GTK_SMPOP))
+    {
+        return false;
+    }
+
+    assert(kind & (GTK_UNOP | GTK_BINOP));
+
+    switch (oper)
```

Comment:
You will need to add quite a few comments to explain these various steps and fliters.
Probably best to add them early to explain to the code reviewers how this method  is working.

This first switch confused me because it obviously only **rejects** a small set of node kinds, 
and I expected this method to be written to **accept** a small set of node kinds
Perhaps you just wanted to eliminate these node at an early stage (for some unknown reason)